src/librustc_target/spec/linux_musl_base.rs

```diff
@@ -21,9 +21,11 @@ pub fn opts() -> TargetOptions {
     //
     // Each target directory for musl has these object files included in it so
     // they'll be included from there.
-    base.pre_link_objects_exe_crt.push("crt1.o".to_string());
-    base.pre_link_objects_exe_crt.push("crti.o".to_string());
-    base.post_link_objects_crt.push("crtn.o".to_string());
+    // FIXME: If musl libc is linked dynamically, gcc will pull its own CRT objects,
+    // but they will be incorrect (at least on glibc host). `-nostdlib` should always
+    // be used always if we want to link correct CRT objects.
```

Comment:
From https://github.com/rust-lang/rust/pull/40113 and related PRs it follows that *not* passing `-nostdlib` and using system CRT objects found by gcc is strongly preferable.

However, that only works if system musl is installed, but for compatibility rustc should be "self-contained" and able to compile for musl target while only having what it brings with itself. (This situation is similar to windows-gnu.)